keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1187 @@
+things like finding order for mutating admission policies that minimizes the
+need for reinvocation. Also, in-process mutation is sufficiently fast,
+especially when compared with webhooks, that it is reasonable to re-run
+mutations to do things like validate that multiple muation policy is still
+applied after all other mutating admission operations have been applied.
+
+### Goals
+
+- Provide an alternative to mutating webhooks for the vast majority of mutating
+  admission use cases.
+- Provide the in-tree extensions needed to build policy frameworks for
+  Kubernetes, again without requiring webhooks for the vast majority of use
+  cases.
+- Provide an out-of-tree implementation of this enhancement (using a webhook)
+  that is supported by the Kubernetes org to provide this enhancement
+  functionality to Kubernetes versions where this enhancement is not available.
+- Provide core functionality as a library so that use cases like GitOps,
+  CI/CD pipelines, and auditing can run the same CEL validation checks
+  that the API server does.
+
+### Non-Goals
+
+- Build a comprehensive in-tree policy framework. We believe the ecosystem is
+  best equipped to explore and develop policy frameworks.
+- Full feature parity with mutating admission webhooks. For example, this
+  enhancement is not expected to ever support making requests to external
+  systems.
+- Replace the admission controllers compiled into the API server. 
+- Static or on-initialization specification of admission config. This is a
+  needed feature but should be solved in a general way and not in this KEP
+  (xref: https://github.com/kubernetes/enhancements/issues/1872).
+
+## Proposal
+
+Introduce `MutatingAdmissionPolicy` API, e.g.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: MutatingAdmissionPolicy
+metadata:
+  name: "sidecar-policy.example.com"
+spec:
+  paramKind:
+    group: mutations.example.com
+    kind: Sidecar
+    version: v1
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE"]
+      resources:   ["pods"]
+  matchConditions:
+    - name: does-not-already-have-sidecar
+      expression: "!object.spec.initContainers.exists(ic, ic.name == params.name)"
+  failurePolicy: Fail
+  mutations:
+    - mutation: >
+        Object{
+          spec: Object.spec{
+            initContainers: [
+              Object.spec.initContainers{
+                name: params.name,
+                image: params.image,
+                args: params.args,
+                restartPolicy: params.restartPolicy
+                // ... other container fields injected here ...
+              }
+            ] + oldObject.spec.initContainers
+          }
+        }
+      mutationValidation:
+        type: Expression
+        expression: "object.spec.initContainers.exists(ic, ic.name == params.name)"
+
+```
+
+The `mutation` field contains a CEL expression that evalutes to a partially
```

Comment:
One question here, does it support like injecting a node label to pod and validate the pod's node labels not conflict. E.g.
pod has a label `foo:bar`, we want to inject a new label like `foo:fuz`, and we found they're conflicted, then fail into pod creation failure.   Thanks.

Reply:
I'm not sure I fully understand the question, but a mutation can add a label, and it can do this conditionally.  A validation can be paired with the mutation to reject requests where there was a conflict.

Reply:
My use case is when a pod is located at a specific namespace, I would like to inject some node selectors and node taints, but before injecting, I would like verify whether they're conflict with pod's current node selectors and taints, if conflict, then reject the pod creation.

Reply:
(I'm going to add a use case for this)